### PR TITLE
[Orc8r] Add upgrade prechecks to deployer tool

### PR DESCRIPTION
**Pull request by karthiksubraveti**
## Summary

Following PR adds upgrade prechecks and post deployment verifications checks to the deployer tool.
Upgrade prechecks
    - Check if terraform state exists
    - Check if deployed EKS version <= EKS version specified in values
    - Check if deployed RDS version <= DB instance version specified in values
    - Check if AWSServiceRoleForAmazonElasticsearchService already exists

Post deployment verifications
    - Helm charts were properly deployed
    - Kubernetes pods are in running state
    - Get all networks and LTE networks from Orc8r and verify if API calls are working as expected
   

For e.g
```
orcl upgrade precheck
..
..
failed: [localhost] (item={'Path': '/aws-service-role/es.amazonaws.com/', 'RoleName': 'AWSServiceRoleForAmazonElasticsearchService', 'RoleId': 'AROA5JG7LAYOZVBCMNO4L', 'Arn': 
...
        "Path": "/aws-service-role/es.amazonaws.com/",
        "RoleId": "AROA5JG7LAYOZVBCMNO4L",
        "RoleName": "AWSServiceRoleForAmazonElasticsearchService"
    },
    "msg": [
        "IAM role AWSServiceRoleForAmazonElasticsearchService already exists",
        "Configure deploy_elasticsearch_service_linked_role to be false",
        "For e.g: orcl configure set -c platform -k deploy_elasticsearch_service_linked_role -v false"
    ]
}
```
## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: orc8r/cloud/deploy/orc8r_deployer/docker/run_deployer [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   echo "Syntax: run_deployer [-d|-r|-b|-h]"
+   echo "options:"
+   echo "-h     Print this Help"
+   echo "-d  deployment dir containing configs and secrets"
+   echo "-r  magma root directory"
+   echo "-b  build the deployer container"
+   echo "note: -deploy_root is mandatory"
+   echo "example: ./run_deployer -deploy_root /tmp/orc8r_14_deployment"
+   echo
+}
```
Comment (andreilee) [inline orc8r/cloud/deploy/orc8r_deployer/docker/run_deployer]: I believe you could avoid the repeated `echo` calls:

https://stackoverflow.com/questions/10969953/how-to-output-a-multiline-string-in-bash

```suggestion
__help="
./run_deployer runs the orc8r deployer container
orc8r deployer contains scripts which enable user to configure, run prechecks
install, upgrade, verify and cleanup an orc8r deployment

Syntax: run_deployer [-d|-r|-b|-h]
options:
-h     Print this Help
-d  deployment dir containing configs and secrets
-r  magma root directory
-b  build the deployer container
note: -deploy_root is mandatory
example: ./run_deployer -deploy_root /tmp/orc8r_14_deployment
"

Print_Help()
{
   echo "$__help"
}
```

File: orc8r/cloud/deploy/orc8r_deployer/docker/run_deployer [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+then
+    echo "Error ${DEPLOY_WORKDIR} not a directory."
+    exit
+fi
+
+MAGMA_ROOT=${PWD/\/orc8r\/cloud\/deploy\/orc8r_deployer\/docker}
+echo "MAGMA ROOT $MAGMA_ROOT"
+
+if $DOCKER_BUILD; then
+docker build --pull -t ksubraveti/orc8r_deployer:latest .
```
Comment (andreilee) [inline orc8r/cloud/deploy/orc8r_deployer/docker/run_deployer]:
```suggestion
    docker build --pull -t ksubraveti/orc8r_deployer:latest .
```

File: orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
+# Copyright 2021 The Magma Authors.
+
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Comment (andreilee) [inline orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py]:
```suggestion
"""
Copyright 2021 The Magma Authors.

This source code is licensed under the BSD-style license found in the
LICENSE file in the root directory of this source tree.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
```

File: orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+import csv
+from subprocess import check_call, CalledProcessError
+from jinja2 import Environment, FileSystemLoader
+from prettytable import PrettyTable
+import click
+from jinja2 import Environment, PackageLoader
+
+def get_input(text):
+    return input(text)
+
```
Comment (andreilee) [inline orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py]: This seems a little redundant to me

File: orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/orcl.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    if not os.path.isdir(constants["secret_dir"]):
+        try:
+            os.makedirs(constants["secret_dir"])
+        except OSError as error:
+            click.echo("failed creating secret directories ", error)
+            sys.exit(1)
+    return constants
+
+# initiialize CLI
```
Comment (andreilee) [inline orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/orcl.py]: typo

File: orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            self.constants['main_tf'],
+            tf_cfgs)
+        render_j2_template(
+            self.constants['tf_dir'],
+            self.constants['vars_tf'],
+            self.tf_vars)
+
+
+    def check(self, component: str) -> bool:
+        ''' check if all mandatory options of a specific component is set '''
```
Comment (andreilee) [inline orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py]: nit
```suggestion
        """ check if all mandatory options of a specific component is set """
```

File: orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py (hunk 4 above)
Comment (karthiksubraveti) [reply orc8r/cloud/deploy/orc8r_deployer/docker/root/scripts/configlib.py]: This is for mocking the get_input in tests.

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Clearing queue. Please make this a draft PR until the "depended" PR is merged.

**Pull request closed without merging**